pydrive2/fs.py

```diff
@@ -0,0 +1,476 @@
+import errno
+import io
+import logging
+import os
+import posixpath
+import threading
+from collections import defaultdict
+
+from fsspec.spec import AbstractFileSystem
+from funcy import cached_property, retry, wrap_prop, wrap_with
+from funcy.py3 import cat
+from tqdm.utils import CallbackIOWrapper
+
+from pydrive2.drive import GoogleDrive
+from pydrive2.fs_utils import IterStream
+
+logger = logging.getLogger(__name__)
+
+FOLDER_MIME_TYPE = "application/vnd.google-apps.folder"
+
+
+def _gdrive_retry(func):
+    def should_retry(exc):
+        from pydrive2.files import ApiRequestError
+
+        if not isinstance(exc, ApiRequestError):
+            return False
+
+        error_code = exc.error.get("code", 0)
+        result = False
+        if 500 <= error_code < 600:
+            result = True
+
+        if error_code == 403:
+            result = exc.GetField("reason") in [
+                "userRateLimitExceeded",
+                "rateLimitExceeded",
+            ]
+        if result:
+            logger.debug(f"Retrying GDrive API call, error: {exc}.")
+
+        return result
+
+    # 16 tries, start at 0.5s, multiply by golden ratio, cap at 20s
+    return retry(
+        16,
+        timeout=lambda a: min(0.5 * 1.618 ** a, 20),
+        filter_errors=should_retry,
+    )(func)
+
+
+class GDriveFileSystem(AbstractFileSystem):
+    def __init__(self, path, google_auth, trash_only=True, **kwargs):
+        self.path = path
+        self.root, self.base = self.split_path(self.path)
+        self.client = GoogleDrive(google_auth)
+        self._trash_only = trash_only
+        super().__init__(**kwargs)
+
+    def split_path(self, path):
+        parts = path.replace("//", "/").rstrip("/").split("/", 1)
+        if len(parts) == 2:
+            return parts
+        else:
+            return parts[0], ""
+
+    @wrap_prop(threading.RLock())
+    @cached_property
+    def _ids_cache(self):
+        cache = {
+            "dirs": defaultdict(list),
+            "ids": {},
+            "root_id": self._get_item_id(
+                self.path,
+                use_cache=False,
+                hint="Confirm the directory exists and you can access it.",
+            ),
+        }
+
+        self._cache_path_id(self.base, cache["root_id"], cache=cache)
+
+        for item in self._gdrive_list(
```

Comment:
it looks very DVC-specific ... should not belong to PyDrive probably

Reply:
> it looks very DVC-specific ... should not belong to PyDrive probably

Can you explain it a bit more? I don't get which part is DVC specific, this only seem to be pre-populating the cache for `root` subdirs.

Reply:
@shcheklein This piece of code is opinionated, but there is no need to touch it right now. Our agreement is to just move fs to pydrive2 with minimal changes.

Reply:
Yep, I understand. I'm worried a bit that we might end up creating some mess in this project only because it makes sense to move things there. The worst case for this project is for this layer being so opinionated that it won't be useful to anyone else, except DVC.

On the other hand, PyDrive2 has some bigger problems to solve, so no strong opinions here. Let's do the best.